src/k2/persistence/plog_client/PlogClient.cpp

```diff
@@ -0,0 +1,203 @@
+PlogClient::init(String clusterName){
+    return _getPersistenceCluster(clusterName);
+}
+
+seastar::future<> 
+PlogClient::_getPlogServerEndpoints() {
+    for(auto& v : _persistenceCluster.persistenceGroupVector){
+        K2INFO("Persistence Group: " << v.name);
+        _persistenceNameMap[v.name] = _persistenceNameList.size();
+        _persistenceNameList.push_back(v.name);
+        
+        std::vector<std::unique_ptr<TXEndpoint>> endpoints;
+        for (auto& url: v.plogServerEndpoints){
+            K2INFO("Plog Server Url: " << url);
+            auto ep = RPC().getTXEndpoint(url);
+            if (ep){
+                endpoints.push_back(std::move(ep));
+            }
+        }
+        if (endpoints.size() == 0){
+            K2INFO("Failed to obtain the Endpoint of Plog Servers");
+            return seastar::make_exception_future<>(std::runtime_error("Failed to obtain the Endpoint of Plog Servers"));
+        }
+        _persistenceMapEndpoints[std::move(v.name)] = std::move(endpoints);
+    }
+    return seastar::make_ready_future<>();
+}
+
+seastar::future<> 
+PlogClient::_getPersistenceCluster(String clusterName){
+    _cpo = CPOClient(String(_cpo_url()));
+    return _cpo.GetPersistenceCluster(Deadline<>(_cpo_timeout()), std::move(clusterName)).
+    then([this] (auto&& result) {
+        auto& [status, response] = result;
+
+        if (!status.is2xxOK()) {
+            K2INFO("Failed to obtain Persistence Cluster" << status);
+            return seastar::make_exception_future<>(std::runtime_error("Failed to obtain Persistence Cluster Map"));
+        }
+
+        _persistenceCluster = std::move(response.cluster);
+        _persistenceMapPointer = rand() % _persistenceCluster.persistenceGroupVector.size();
+        _persistenceMapEndpoints.clear();
+        return _getPlogServerEndpoints();
+    });
+}
+
+// TODO: If the create call fails, we should try and create the plog in another persistence group.
+seastar::future<std::tuple<Status, String>> PlogClient::create(){
+    String plogId = _generatePlogId();
+    dto::PlogCreateRequest request{.plogId = plogId};
+    
+    std::vector<seastar::future<std::tuple<Status, dto::PlogCreateResponse> > > createFutures;
+    for (auto& ep:_persistenceMapEndpoints[_persistenceNameList[_persistenceMapPointer]]){
+        createFutures.push_back(RPC().callRPC<dto::PlogCreateRequest, dto::PlogCreateResponse>(dto::Verbs::PERSISTENT_CREATE, request, *ep, _plog_timeout()));
+    }
+    return seastar::when_all_succeed(createFutures.begin(), createFutures.end())
+        .then([this, plogId](std::vector<std::tuple<Status, dto::PlogCreateResponse> >&& results) { 
+            Status return_status;
+            for (auto& result: results){
+                auto& [status, response] = result;
+                return_status = std::move(status);
+                if (!return_status.is2xxOK()) 
+                    break;
+            }
+            if (return_status.code == 530){

```

Comment:
I think this code is now 409 conflict.
Also, you have an infinite loop possibility (e.g. if the server has a bug and keeps returning 409). we can do it later, but for now at least put a TODO that we have to stop retrying after some number of retries.

Reply:
added the retry